### PR TITLE
Changes to improve GE0  GEMSegments

### PR DESCRIPTION
**Pull request by watson-ij**
#### PR description:

These changes allow GE0 GEMSegments in GEM to act the same as the current ME0Segment:
- Add `deltaPhi` to GEMSegments to match the ME0Segment
- Copied the deltaPhi calculator to GEMSuperChamber from ME0Chamber (which is actually the equivalent of GEMSuperChamber)
- Adapted ME0SegAlgRU to GE0SegAlgRU so the RU segment finder can run from GEM RecHits (for now, keep default GEMSegmentBuilder, but we may later want to switch to GE0SegAlgRU, or split the algorithm used for GE0 and GE[12]/1)

@jshlee @dildick Since we are updating the GEMSegment with deltaPhi and therefore changing the GEMSegment interface anyway, I went ahead and removed the timing information from GEMSegments. The GEM only outputs bunch crossing information, and from what I've been told there's no plan to include timing, so I think this info is unused and not planned to be used in the future. Let me know if you prefer to reverse these changes.

#### PR validation:

I checked runTheMatrix 24821.0 (TenMuE) with the GE0Geometry in Geometry/GEMGeometry up to DIGI + GEM RecoHit/Segment reconstruction. Segments look to be made correctly, all stations (including GE0) are producing them with the current setup.
I also checked the GE0SegAlgRU produces segments, but we should compare the ME0Segments and GE0Segments produced later, once the whole pipeline is working.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by dildick** [comment]
@slava77 @perrotta, Ian is proposing to remove 2 members from GEMSegment. Do we need an ioread rule for this?

File: Geometry/GEMGeometry/src/GEMSuperChamber.cc [hunk 1]
```diff
@@ -37,3 +37,26 @@ const GEMChamber* GEMSuperChamber::chamber(int isl) const {
   }
   return nullptr;
 }
+
+float GEMSuperChamber::computeDeltaPhi(const LocalPoint& position, const LocalVector& direction) const {
+  auto extrap = [](const LocalPoint& point, const LocalVector& dir, double extZ) -> LocalPoint {
+    double extX = point.x() + extZ * dir.x() / dir.z();
```
Comment (cvuosalo) [inline Geometry/GEMGeometry/src/GEMSuperChamber.cc:43]: Please prevent division by zero by checking `dir.z()`.

**Comment by cvuosalo** [comment]
Please explain the comparison differences.

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+  }
+}
+
+void GE0SegAlgoRU::increaseProtoSegment(const float maxChi2,
+                                        std::unique_ptr<MuonSegFit>& current_fit,
+                                        HitAndPositionPtrContainer& current_proto_segment,
+                                        const HitAndPosition& new_hit) const {
+  HitAndPositionPtrContainer new_proto_segment = current_proto_segment;
+  auto new_fit = addHit(new_proto_segment, new_hit);
+  //	edm::LogVerbatim("GE0SegAlgoRU") << "[GE0SegAlgoRU::increaseProtoSegment] new chi2 = "<<new_fit->chi2()<<" new chi2/dof = "<<new_fit->chi2()/new_fit->ndof() <<" ==> add: " << (new_fit->chi2()/new_fit->ndof() < maxChi2 ) <<std::endl;
```
Comment (kpedro88) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:574]: delete commented-out code

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.h [hunk 3]
```diff
@@ -0,0 +1,157 @@
+#ifndef GEMSegment_GE0SegAlgoRU_h
```
Comment (kpedro88) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.h:1]: include guard should have subsystem name

File: RecoLocalMuon/GEMSegment/python/gemSegments_cfi.py [hunk 4]
```diff
@@ -1,10 +1,25 @@
...
+    maxPhiAdditional = cms.double(0.001096605744), #Assuming 384 strips
+    maxETASeeds = cms.double(0.1), #Assuming 8 eta partitions
+    maxTOFDiff = cms.double(25),
+    requireCentralBX = cms.bool(True), #require that a majority of hits come from central BX
+    minNumberOfHits = cms.uint32(4),
+)
+
 gemSegments = cms.EDProducer("GEMSegmentProducer",
     gemRecHitLabel = cms.InputTag("gemRecHits"),
+    # algo_name = cms.string("GE0SegAlgoRU"),
```
Comment (kpedro88) [inline RecoLocalMuon/GEMSegment/python/gemSegments_cfi.py:19]: delete commented-out code

**Comment by kpedro88** [comment]
@cvuosalo I understand the comparison differences: the time members were removed from the data format. The object still exists, so the reco comparison code tries to fill the histogram, but the relevant member isn't there, so it fills with no/default value.

File: DataFormats/GEMRecHit/interface/GEMSegment.h [hunk 5]
```diff
@@ -93,9 +94,8 @@ class GEMSegment final : public RecSegment {
   LocalVector theLocalDirection;    // in chamber frame - the GeomDet local coordinate system
   AlgebraicSymMatrix theCovMatrix;  // the covariance matrix
   double theChi2;                   // the Chi squared of the segment fit
-  double theTimeValue;              // the best time estimate of the segment
-  double theTimeUncrt;              // the uncertainty on the time estimation
   float theBX;                      // the bunch crossing
+  float theDeltaPhi;                // Difference in segment phi position: outer layer - inner lay
```
[the remark this answers is not in the document]
Comment (watson-ij) [reply DataFormats/GEMRecHit/interface/GEMSegment.h:98]: @kpedro88 Thanks for the comments. For here, based on comparing some other classes, I added a ioread for older versions of GEMSegment without theDeltaPhi. Do I need to also do something for the dropped timing information, or these will just be ignored?

**Comment by watson-ij** [comment]
Thanks for the comments, I'm going through cleaning up the GE0SegAlgRU code now; this was inherited from ME0SegAlgRU, but I'll leave that one as is, since we should eventually be able to just remove it. For the differences, the histograms also show these fields only ever get filled with zeros.

**Comment by slava77** [comment]
@cmsbuild please test

File: DataFormats/GEMRecHit/interface/GEMSegment.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (kpedro88) [reply DataFormats/GEMRecHit/interface/GEMSegment.h:98]: Dropped information will just be ignored.

File: DataFormats/GEMRecHit/interface/GEMSegment.h [hunk 6]
```diff
@@ -40,8 +40,8 @@ class GEMSegment final : public RecSegment {
              const LocalVector& direction,
              const AlgebraicSymMatrix& errors,
              double chi2,
```
Comment (perrotta) [inline DataFormats/GEMRecHit/interface/GEMSegment.h:42]: Now you pass everything as `float`: is it still necessary to pass chi2 as a `double`?

File: DataFormats/GEMRecHit/interface/GEMSegment.h [hunk 7]
```diff
@@ -93,9 +94,8 @@ class GEMSegment final : public RecSegment {
   LocalVector theLocalDirection;    // in chamber frame - the GeomDet local coordinate system
   AlgebraicSymMatrix theCovMatrix;  // the covariance matrix
   double theChi2;                   // the Chi squared of the segment fit
```
Comment (perrotta) [inline DataFormats/GEMRecHit/interface/GEMSegment.h:96]: ditto

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+}
+
+bool GE0SegAlgoRU::isHitNearSegment(const float maxETA,
+                                    const float maxPHI,
+                                    const std::unique_ptr<MuonSegFit>& fit,
+                                    const HitAndPositionPtrContainer& proto_segment,
+                                    const HitAndPosition& h) const {
+  //Get average eta, based on the two seeds...asssumes that we have not started pruning yet!
+  const float avgETA = (proto_segment[1]->gp.eta() + proto_segment[0]->gp.eta()) / 2.;
+  //	edm::LogVerbatim("GE0SegAlgoRU") << "[GE0SegAlgoRU::isHitNearSegment] average eta = "<<avgETA<<" additionalHit eta = "<<h.gp.eta() <<" ==> dEta = "<<std::fabs(h.gp.eta() - avgETA) <<" ==> return "<<(std::fabs(h.gp.eta() - avgETA) < std::max(maxETA,0.01f ))<<std::endl;
```
Comment (perrotta) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:431]: `std::abs` (even though I think you can remove the commented out part)

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 9]
```diff
@@ -0,0 +1,579 @@
...
+  for (auto i1 = ib; i1 != ie; ++i1) {
+    const auto& h1 = *i1;
+
+    //skip if rh is used and the layer tat has big rh multiplicity(>25RHs)
+    if (used[h1.idx])
+      continue;
+    if (recHits_per_layer[h1.layer - 1] > 100)
+      continue;
+
+    //    	bool segok = false;
```
Comment (perrotta) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:220]: The commented out references to `segok` can be removed (here and below)

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 10]
```diff
@@ -0,0 +1,579 @@
...
+  const float avgETA = (proto_segment[1]->gp.eta() + proto_segment[0]->gp.eta()) / 2.;
+  //	edm::LogVerbatim("GE0SegAlgoRU") << "[GE0SegAlgoRU::isHitNearSegment] average eta = "<<avgETA<<" additionalHit eta = "<<h.gp.eta() <<" ==> dEta = "<<std::fabs(h.gp.eta() - avgETA) <<" ==> return "<<(std::fabs(h.gp.eta() - avgETA) < std::max(maxETA,0.01f ))<<std::endl;
+  if (std::fabs(h.gp.eta() - avgETA) > std::max(maxETA, 0.01f))
+    return false;
+
+  //Now check the dPhi based on the segment fit
+  GlobalPoint globIntercept = globalAtZ(fit, h.lp.z());
+  float dPhi = deltaPhi(h.gp.barePhi(), globIntercept.phi());
+  //check to see if it is inbetween the two rolls of the outer and inner hits
+  //	edm::LogVerbatim("GE0SegAlgoRU") << "[GE0SegAlgoRU::isHitNearSegment] projected phi = "<<globIntercept.phi()<<" additionalHit phi = "<<h.gp.barePhi() <<" ==> dPhi = "<<dPhi <<" ==> return "<<(std::fabs(dPhi) <  std::max(maxPHI,0.001f))<<std::endl;
```
Comment (perrotta) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:439]: `std::abs` (even though I think you can remove the commented out part)

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 11]
```diff
@@ -0,0 +1,579 @@
...
+  return (std::fabs(dPhi) < std::max(maxPHI, 0.001f));
```
Comment (perrotta) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:440]: `std::abs`

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 12]
```diff
@@ -0,0 +1,579 @@
...
+bool GE0SegAlgoRU::areHitsConsistentInTime(const float maxTOF,
+                                           const HitAndPositionPtrContainer& proto_segment,
+                                           const HitAndPosition& h) const {
+  std::vector<float> tofs;
+  tofs.reserve(proto_segment.size() + 1);
+  tofs.push_back(h.rh->BunchX());
+  for (const auto* ih : proto_segment)
+    tofs.push_back(ih->rh->BunchX());
+  std::sort(tofs.begin(), tofs.end());
+  if (std::fabs(tofs.front() - tofs.back()) < maxTOF + 1.0)
```
Comment (perrotta) [inline RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:452]: `std::abs` (but please look for fabs everywhere, we quite likely missed some)

**Review by perrotta — COMMENTED** [review]
First set of comments to the code, after a quick glance

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+
+  auto doStd = [&]() {
+    for (unsigned int n_seg_min = 6u; n_seg_min >= stdParameters.minNumberOfHits; --n_seg_min)
+      lookForSegments(stdParameters, n_seg_min, rechits, recHits_per_layer, used, segments);
+  };
+  auto doDisplaced = [&]() {
+    for (unsigned int n_seg_min = 6u; n_seg_min >= displacedParameters.minNumberOfHits; --n_seg_min)
+      lookForSegments(displacedParameters, n_seg_min, rechits, recHits_per_layer, used, segments);
+  };
+  // Not currently used
```
[the remark this answers is not in the document]
Comment (watson-ij) [reply RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:150]: @kpedro88 we will have a student check the GE0 vs ME0 setup soon, and we are also planning to have them test the various options after the GE0 validation. There is a wide switch, but the code as written would switch various paths that I would prefer to just keep off until we understand all the effects it would have.

File: DataFormats/GEMRecHit/interface/GEMSegment.h (hunk 6 above)
Comment (watson-ij) [reply DataFormats/GEMRecHit/interface/GEMSegment.h:42]: @perrotta I think this was originally to match the interface of CSCSegment, which also defines theChi2 as double. Is it prefered to just switch to float here?

**Comment by watson-ij** [comment]
The updates should address the comments of @kpedro and @perrotta, I also removed some additional TOF code from GE0SegAlgoRU

File: DataFormats/GEMRecHit/interface/GEMSegment.h (hunk 6 above)
Comment (perrotta) [reply DataFormats/GEMRecHit/interface/GEMSegment.h:42]: If using a double is any useful in the code it can stay as such. If not, better to reduce precision (and also make it uniform with the other similar parameters)

**Comment by perrotta** [comment]
please test

File: RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc [hunk 14]
```diff
@@ -100,25 +100,25 @@ std::vector<GEMSegment> GE0SegAlgoRU::run(const GEMSuperChamber* chamber, const
...
                                      << " Glb y = " << std::showpos << std::setw(9) << rhLP.y()
-                                     << " Time = " << std::showpos << rechits[iH].rh->BunchX() << " -- "
-                                     << ge0id.rawId() << " = " << ge0id << " ]" << std::endl;
+                                     << " Time = " << std::showpos << h.rh->BunchX() << " -- " << ge0id.rawId() << " = "
+                                     << ge0id << " ]" << std::endl;
   }
 #endif
 
-  if (rechits.size() < 3 || rechits.size() > 300) {
+  if (rechits.size() < stdParameters.minNumberOfHits || rechits.size() > stdParameters.maxNumberOfHits) {
```
[the remark this answers is not in the document]
Comment (watson-ij) [reply RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc:113]: Yes, but I think if there are less hits than the minNumberOfHits we require for a segment, we would anyway reject all protosegments, so it shouldn't change the result and gets rid of the magic number.

File: DataFormats/GEMRecHit/interface/GEMSegment.h (hunk 6 above)
Comment (watson-ij) [reply DataFormats/GEMRecHit/interface/GEMSegment.h:42]: In GEMCSCSegment, we combine the chi2 from CSC and GEM, so I would have thought it would be better to keep it the same precision in this case. @jshlee do you have an opinion?

**Comment by perrotta** [comment]
@watson-ij : is there any plan to get rid of the similar ME0 code once this PR will get merged? If so, why not to do it right now here?

**Comment by watson-ij** [comment]
> @watson-ij : is there any plan to get rid of the similar ME0 code once this PR will get merged? If so, why not to do it right now here?

Yes, we plan to get rid of this code, but the ME0 code will still be needed to run upgrade simulation until the full GE0 replacement is done. The idea is to change over (incrementally), and remove ME0 once GE0 is fully in and working. For the other PR  #31481, that code is (as far as we could see) not in use anywhere, so should be safe to remove immediately.

**Comment by kpedro88** [comment]
A few comments were missed:
https://github.com/cms-sw/cmssw/pull/31480/files#r489797636
https://github.com/cms-sw/cmssw/pull/31480/files#r489797650
https://github.com/cms-sw/cmssw/pull/31480/files#r489802337

File: Geometry/GEMGeometry/src/GEMSuperChamber.cc (hunk 1 above)
Comment (cvuosalo) [reply Geometry/GEMGeometry/src/GEMSuperChamber.cc:45]: @watson-ij Please address this issue.

Comment (kpedro88) [reply Geometry/GEMGeometry/src/GEMSuperChamber.cc:45]: If you look at the latest version of the code, there's a check for z==0 above this line

Comment (cvuosalo) [reply Geometry/GEMGeometry/src/GEMSuperChamber.cc:45]: Sorry, must be going blind. ;-).  Looks good.

**Comment by cvuosalo** [comment]
+1

**Comment by watson-ij** [comment]
> A few comments were missed:
> https://github.com/cms-sw/cmssw/pull/31480/files#r489797636
> https://github.com/cms-sw/cmssw/pull/31480/files#r489797650
> https://github.com/cms-sw/cmssw/pull/31480/files#r489802337

Yep, sorry, those should be taken care of now.

**Comment by perrotta** [comment]
please test

**Comment by watson-ij** [comment]
Sorry, I realized missed one -> to . change in the debug mode code. It compiles now with and without -DEDM_ML_DEBUG.

**Comment by watson-ij** [comment]
I just chatted with @jshlee and he agrees that using float for chi2 is preferable, so I also made that change. I think that was the last outstanding issue.

**Comment by watson-ij** [comment]
Actually, no, I spoke too soon, the reason we are using double there is that we inherit from RecSegment which requires double chi2 in the interface. I've reverted the changes

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
code-checks

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
> Actually, no, I spoke too soon, the reason we are using double there is that we inherit from RecSegment which requires double chi2 in the interface. I've reverted the changes

I'm convinced that it could be more conveniently migrated everywhere, thus avoiding type conversions back and forth from `float` to `double...` But I agree that it is not something that has to be dealt with in this PR

**Comment by kpedro88** [comment]
+upgrade

**Comment by cvuosalo** [comment]
+1

**Comment by perrotta** [comment]
+1
- It adapts GEMSegments to allow GE0 in them
- A new GE0SegAlgRU plugin is also introduced here: it will eventually replace the ME0SegAlgRU one, but it is not yet inserted by default in the Phase2 pipeline
  - Some code overlap with ME0SegAlgRU is unavoidable: it will become irrelevant once ME0SegAlgRU will be dropped
- Jenkins tests pass and show only differences for the removed time variable that was still there in the previous GEMSegment
- GE0SegAlgRU not tested with the central tests because not yet included in the default reco: a validation has been performed by the authors, and reported in the PR description

File: RecoLocalMuon/GEMSegment/python/gemSegments_cfi.py [hunk 15]
```diff
@@ -1,10 +1,26 @@
...
+    maxPhiSeeds = cms.double(0.001096605744), #Assuming 384 strips
+    maxPhiAdditional = cms.double(0.001096605744), #Assuming 384 strips
+    maxETASeeds = cms.double(0.1), #Assuming 8 eta partitions
+    maxTOFDiff = cms.double(25),
+    requireCentralBX = cms.bool(True), #require that a majority of hits come from central BX
+    minNumberOfHits = cms.uint32(4),
+    maxNumberOfHits = cms.uint32(300),
+    maxNumberOfHitsPerLayer = cms.uint32(100),
+)
+
```
Comment (silviodonato) [inline RecoLocalMuon/GEMSegment/python/gemSegments_cfi.py:19]: Why did you add `RU_GE0` in `gemSegments_cfi.py`? This file should contain only the default values of `gemSegments`

Comment (watson-ij) [reply RecoLocalMuon/GEMSegment/python/gemSegments_cfi.py:19]: Because this is the default for the RU algorithm, which is the default reconstruction for ME0. For GE0, we will have a student check the various segment algorithms and possibly add the option to change algorithm by station, so I wanted to have this here to make it easy for them to switch algorithms.

**Comment by silviodonato** [comment]
+1
